design-proposals/vhostuser/vhostuser-interface.md

```diff
@@ -0,0 +1,166 @@
+- There should not be any scalability issues, as the feature is to create an interface when useful.
+- VM migration works the same way, as the only change is having a new interace, which only requires a dataplane of the host in userpsace.
+
+## Update/Rollback Compatibility
+- Should have no impact.
+
+## Functional Testing Approach
+Functional test can:
+
+- Create VM with vhostUser interface
+- Create VM with multiple interface types. (Bridge + vhostuser type)
+
+# Implementation Phases
+- The design will be implemented in two phases:
+    - Add DPDK support on kubevirt/kubevirtci
+        - Enable OVS with DPDK
+        - Enable an option in gocli for DPDK support
+        - Add UserSpace CNI in the setup instead of or along with calico.
+    - Add vhostUser Interface type in kubevirt/kubevirt
+        - Add vHostUser Interface type
+        - Create Pod template to support the Interface
+        - Create appropriate virsh xml elements
+        - Add E2E test to send traffic between the 2 VMs created.
+
+## Annex
+
+
+A sample DownwardAPi mght look this way:
+
+```yaml
+k8s.v1.cni.cncf.io/network-status="[{\n    \"name\": \"k8s-kubemanager-kubernetes-CNI/default-podnetwork\",\n    \"interface\": \"eth0\",\n    \"ips\": [\n        \"10.244.2.4\"\n    ],\n    \"mac\": \"02:d7:b3:27:b3:b2\",\n    \"default\": true,\n    \"dns\": {}\n},{\n    \"name\": \"kubevirttest/vn-blue\",\n    \"interface\": \"net1\",\n    \"ips\": [\n        \"19.1.1.2\"\n    ],\n    \"mac\": \"02:c6:da:d3:ab:07\",\n    \"dns\": {},\n    \"device-info\": {\n        \"type\": \"vhost-user\",\n        \"vhost-user\": {\n            \"mode\": \"server\",\n            \"path\": \"2ea1931c-2935-net1\"\n        }\n    }\n}]"
+k8s.v1.cni.cncf.io/networks="[{\"interface\":\"net1\",\"name\":\"vn-blue\",\"namespace\":\"kubevirttest\",\"cni-args\":{\"interface-type\":\"virtio\"}}]"
+kubectl.kubernetes.io/default-container="compute"
+kubernetes.io/config.seen="2023-05-16T16:50:04.224937236Z"
+kubernetes.io/config.source="api"
+kubevirt.io/domain="vm-single-virtio"
+kubevirt.io/migrationTransportUnix="true"
+post.hook.backup.velero.io/command="[\"/usr/bin/virt-freezer\", \"--unfreeze\", \"--name\", \"vm-single-virtio\", \"--namespace\", \"kubevirttest\"]"
+post.hook.backup.velero.io/container="compute"
+pre.hook.backup.velero.io/command="[\"/usr/bin/virt-freezer\", \"--freeze\", \"--name\", \"vm-single-virtio\", \"--namespace\", \"kubevirttest\"]"
+pre.hook.backup.velero.io/container="compute"
+```
+The CNI should update the /ect/podnetinfo/annotaions with something similar, to update the annotations of the virt-launcher pod.
+
+The NAD can be generic which can just be used for defining the networks. The below is a NAD definition from userspace CNI based on ovs-dpdk by intel.
+
+```yaml
+apiVersion: "k8s.cni.cncf.io/v1"
+kind: NetworkAttachmentDefinition
+metadata:
+  name: userspace-ovs-net-1
+spec:
+  config: '{
+        "cniVersion": "0.3.1",
+        "type": "userspace",
```

Comment:
Why is it called "userspace" and not "vhost-user-net"? Is the idea to support other userspace network interfaces in the future?

Reply:
this must have the name of the CNI binary which will run on the host. 

AFAIU, the "reference" CNI to use is userspace CNI, whose binary is named `userspace`: https://github.com/intel/userspace-cni-network-plugin#work-standalone

Reply:
@maiqueb @nvinnakota10 
Thanks for sharing the link to the Userspace CNI plugin! It seems the Userspace CNI plugin doesn't define how Pods need to be configured - it's left up to the user to create volumes, choose paths, and I'm not sure it uses annotations and the downward API.

I think a standard interface should be defined by the Userspace CNI plugin, not in KubeVirt. Then other container workloads can benefit from a standard interface that doesn't require coming up with configuration (volumes, paths, etc) from scratch.

The KubeVirt proposal would just be about implementing the Userspace CNI interface, not defining annotations, etc.

Reply:
As a first step towards this, I suggest splitting this proposal into two parts:
1. The standard Userspace CNI plugin interface that allows any Pod to detect vhost-user-net devices and consume them. This will eventually go into the Userspace CNI plugin project (and other CNI plugins can implement a compatible interface).
2. The KubeVirt-specific part that describes how to implement the interface in virt-launcher, etc.